### PR TITLE
ARM64 preperations

### PR DESCRIPTION
**Pull request by undingen**
This PR refactors the JIT to make it easy to add support for ARM64.
Everything except the IC code is converted to use the newly introduced helper functions (used functions where possible and a few macros for branches where we have to). The generated code is slightly different for `STORE_FAST` and `DELETE_FAST` bytecodes and I fixed a bug in `END_FINALLY`.

I still have to cleanup the real arm implementations so this does not contain any of it yet but I hope to have it ready by next week (without the ICs). And plan on fixing the IC stuff afterwards.

It's best to look at every commit individually.

### CODE REVIEW

File: Python/aot_ceval_jit.c [hunk 1]
```diff
@@ -3215,15 +3215,22 @@ void* jit_func(PyCodeObject* co, PyThreadState* tstate) {
...
 #endif
 
     size_t size;
-    dasm_link(Dst, &size);
+    int dasm_link_err = dasm_link(Dst, &size);
+    if (dasm_link_err) {
+#if JIT_DEBUG
+        fprintf(stderr, "dynasm_link() returned error %x", dasm_link_err);
+        JIT_ASSERT(0, "");
+#endif
```
Comment (kmod) [inline Python/aot_ceval_jit.c:3361]: Just an idea -- if you extend JIT_ASSERT like we did with the old [RELEASE_ASSERT](https://github.com/pyston/pyston_v1/blob/v1.0/src/core/common.h#L53) then these four lines could just be `JIT_ASSERT(dasm_link_err == 0, "dynasm_link() returned error %x", dasm_link_err)`

File: Python/aot_ceval_jit.c [hunk 2]
```diff
@@ -921,20 +1002,19 @@ static void emit_decref(Jit* Dst, int r_idx, int preserve_res) {
     emit_dec_qword_ptr(Dst, &_Py_RefTotal, 1 /* can_use_tmp_reg  */);
 #endif
     | add qword [Rq(r_idx)], -1
-
     // normally we emit the dealloc call into the cold section but if we are already inside it
     // we have to instead emit it inline
     int use_inline_decref = Dst->current_section == SECTION_COLD;
     if (use_inline_decref) {
-        | jnz >8
+        | branch_ne >8 // this is same as 'branch if not zero'
```
Comment (kmod) [inline Python/aot_ceval_jit.c]: I always have to check whether ne is nz or z, what do you think about having a `branch_nz` macro as well?

**Review by kmod — APPROVED** [review]
Looks great!

**Comment by undingen** [comment]
good idea!

**Pull request merged**